File: php/src/Level4.php
<?php declare(strict_types=1);

namespace DJB;

class Level4 extends Level
{
    public function __construct()
    {
        parent::__construct(4);
    }

    public function run()
    {
        $cars = [];
        foreach($this->input->cars as $car) {
            $cars[$car->id] = $car;
        }
        
        $rentals = [];
        foreach($this->input->rentals as $rental) {
            $car = $cars[$rental->car_id];
            $duration = (new \DateTime($rental->end_date))->diff(new \DateTime($rental->start_date))->format("%a") + 1;
            
            $price = $rental->distance * $car->price_per_km + $this->durationPrice($duration, $car->price_per_day);
            
            $commission = $this->getCommission($price, $duration);
            $rentals[] = (object)[
                'id' => $rental->id,
                'actions' => $this->getActions($price, $commission)
            ];
        }

        $output = (object)[
            'rentals' => $rentals
        ];

        return $output;       
    }

    private function getActions(int $price, $commission)
    {
        return [
            (object)[
                'who' => 'driver',
                'type' => 'debit',
                'amount' => $price
            ],
            (object)[
                'who' => 'owner',
                'type' => 'credit',
                'amount' => $price - $commission->insurance_fee - $commission->assistance_fee - $commission->drivy_fee
            ],
            (object)[
                'who' => 'insurance',
                'type' => 'credit',
                'amount' => $commission->insurance_fee
            ],
            (object)[
                'who' => 'assistance',
                'type' => 'credit',
                'amount' => $commission->assistance_fee
            ],
            (object)[
                'who' => 'drivy',
                'type' => 'credit',
                'amount' => $commission->drivy_fee
            ]                        
        ];
    }

    private function getCommission(int $price, int $duration)
    {
        $price *= 0.3;
        return (object)[
            'insurance_fee' => (int)($price * 0.5),
            'assistance_fee' => $duration * 100,
            'drivy_fee' => (int)($price * 0.5) - $duration * 100
        ];
    }

    private function durationPrice(int $duration, int $price_per_day)
    {
        $steps = [
            1 => 10,
            4 => 30,
            10 => 50
        ];

        krsort($steps);

        $price = 0;
        foreach($steps as $days => $reduction) {
            if($duration <= $days) {
                continue;
            }

            $price += ($duration - $days) * $price_per_day * (100 - $reduction) / 100;
            $duration -= ($duration - $days);

        }
        
        return $price + $duration * $price_per_day;
    }
}

File: elixir/lib/level4.ex
defmodule Drivy.Level4 do
  import Drivy

  def run(input) do
    cars = to_map(input["cars"])
    rentals = input["rentals"]

    rentals =
      Enum.map(rentals, fn r ->
        car = Map.get(cars, r["car_id"])
        duration = to_days(r["start_date"], r["end_date"])
        price = get_price(car, r)
        commission = get_commission(price, duration)

        %{
          "id" => r["id"],
          "actions" => get_actions(price, commission)
        }
      end)

    %{"rentals" => rentals}
  end

  def get_commission(price, duration) do
    price = price * 0.3

    %{
      "insurance_fee" => Float.ceil(price * 0.5) |> round,
      "assistance_fee" => Float.ceil(duration * 100.0) |> round,
      "drivy_fee" => Float.ceil(price * 0.5 - duration * 100.0) |> round
    }
  end

  defp get_actions(price, commission) do
    [
      %{
        "who" => "driver",
        "type" => "debit",
        "amount" => price
      },
      %{
        "who" => "owner",
        "type" => "credit",
        "amount" =>
          price - commission["insurance_fee"] - commission["assistance_fee"] -
            commission["drivy_fee"]
      },
      %{
        "who" => "insurance",
        "type" => "credit",
        "amount" => commission["insurance_fee"]
      },
      %{
        "who" => "assistance",
        "type" => "credit",
        "amount" => commission["assistance_fee"]
      },
      %{
        "who" => "drivy",
        "type" => "credit",
        "amount" => commission["drivy_fee"]
      }
    ]
  end

  defp get_price(car, rental) do
    duration = to_days(rental["start_date"], rental["end_date"])
    car["price_per_km"] * rental["distance"] + get_reduction(duration, car["price_per_day"])
  end

  def get_reduction(duration, price_per_day) do
    cond do
      duration > 10 -> (duration - 10) * price_per_day * 0.5 + get_reduction(10, price_per_day)
      duration > 4 -> (duration - 4) * price_per_day * 0.7 + get_reduction(4, price_per_day)
      duration > 1 -> (duration - 1) * price_per_day * 0.9 + get_reduction(1, price_per_day)
      true -> duration * price_per_day
    end
  end
end
